src/test/random/test_random.rs

```diff
@@ -0,0 +1,119 @@
+
+fn main() -> Result<(), String> {
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let tests: Vec<test_utils::ShadowTest<_, _>> = vec![
+        test_utils::ShadowTest::new("test_dev_urandom", test_dev_urandom, set![TestEnv::Libc]),
+        test_utils::ShadowTest::new("test_dev_random", test_dev_random, set![TestEnv::Libc]),
+        test_utils::ShadowTest::new("test_rand", test_rand, set![TestEnv::Libc]),
+        test_utils::ShadowTest::new("test_getrandom", test_getrandom, set![TestEnv::Libc]),
+    ];
+
+    test_utils::run_tests(&tests, summarize)?;
+
+    println!("Success.");
+    Ok(())
+}
+
+fn check_randomness(fracs: &[f64], len: usize) -> Result<(), String> {
+    let mut buckets = [0_u8; BUCKETLEN];
+
+    for i in 0..len {
+        let percent = (fracs[i] * 100_f64) as u8;
+        assert!(
+            percent <= 100,
+            format!("invalid random percent value: {}", percent)
+        );
+        let j = percent as usize % BUCKETLEN;
+        buckets[j] += 1;
+    }
+
+    let mut fail = false;
+    println!("bucket values:");
+    for i in 0..BUCKETLEN {
+        println!("bucket[{}] = {}", i, buckets[i]);
+        if buckets[i] == 0 {
+            fail = true;
+        }
+    }
+
+    if fail {
+        let err_msg = "failed to get random values across entire range";
+        println!("{}", &err_msg);
+        return Err(err_msg.to_string());
+    } else {
+        return Ok(());
+    }
+}
+
+fn test_path_helper(path: &'static str) -> Result<(), String> {
+    use std::io::Read;
+
+    let mut file =
+        std::fs::File::open(path).map_err(|e| format!("error: cannot open file: {:?}", e))?;
+
+    let mut values = [0_f64; RGENLEN];
+
+    for i in 0..RGENLEN {
+        let mut rv = [0_u8; 4];
+        file.read_exact(&mut rv)
+            .map_err(|_| "error reading file".to_string())?;
+        values[i] = u32::from_be_bytes([rv[0], rv[1], rv[2], rv[3]]) as f64 / core::u32::MAX as f64;
+    }
+
+    check_randomness(&values, RGENLEN)
+}
+
+fn test_dev_urandom() -> Result<(), String> {
+    test_path_helper("/dev/urandom")
+}
+
+fn test_dev_random() -> Result<(), String> {
+    test_path_helper("/dev/random")
+}
+
+fn test_rand() -> Result<(), String> {
+    let mut values = [0_f64; RGENLEN];
+
+    for i in 0..RGENLEN {
+        let random_value = unsafe { libc::rand() };
+
+        if random_value < 0 || random_value > libc::RAND_MAX {
+            let err_msg = "error: rand returned bytes outside of expected range";
+            println!("{}", &err_msg);
+            return Err(err_msg.to_string());
+        }
+
+        values[i] = random_value as f64 / libc::RAND_MAX as f64;
+    }
+
+    check_randomness(&values, RGENLEN)
+}
+
+fn test_getrandom() -> Result<(), String> {
+    let mut values = [0_f64; RGENLEN];
+
+    for i in 0..RGENLEN {
+        let mut rv = [0_u8; 4];
+
+        if unsafe { libc::getrandom(rv.as_mut_ptr() as *mut libc::c_void, rv.len(), 0) } <= 0 {
```

Comment:
There is a linking problem with getrandom, since some of our supported platforms use older glibc versions without this wrapper. You'll need to make the syscall directly, as the comments say in the C code.